lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+const URL_FILTERS = [
+  (item) => !!item.url,
+  (item) => !!(new URL(item.url)),
+  (item) => ALLOWED_PROTOCOLS.has(new URL(item.url).protocol),
+  (item) => !DISALLOWED_HOSTS.has(new URL(item.url).hostname)
+];
+
+Cu.importGlobalProperties(["fetch"]);
+Cu.importGlobalProperties(["URL"]);
+Cu.import("resource://gre/modules/Task.jsm");
+
+function PreviewProvider() {
+  this.init();
+}
+
+PreviewProvider.prototype = {
+  _timeoutID: null,
+
+  /**
+    * Clean up cache on addon bootup
+    */
+  cleanUpCache: Task.async(function*() {
+    let currentTime = Date.now();
+    Object.keys(ss.storage.embedlyData).forEach(item => {
+      if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+        delete ss.storage.embedlyData[item];
+      }
+    });
+    this._cacheTimeout();
+  }),
+
+  /**
+    * Set up the timer for the cache to be cleaned every 24 hours
+    */
+  _cacheTimeout() {
+    this._timeoutID = setTimeout(() => {
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  },
+
+  /**
+    * Filter out sites that do not have acceptable protocols and hosts
+    */
+  _URLFilter(definition) {
+    return function(item) {
+      let result = true;
+      definition.forEach((test) => {
+        if (!test(item)) {
+          result = false;
+        }
+      });
+      return result;
+    };
+  },
+
+  /**
+    * Santize the URL to remove any unwanted or sensitive information about the link
+    */
+  _sanitizeURL(site) {
+    if (!site) {
+      return "";
+    }
+
+    let queryParams = new Map();
+    let safeQueryParam = "";
+    let url = new URL(site);
+
+    if (url.search.length === 0) {
+      return url.toString();
+    }
+    // extract and parse the query parameters, if any
+    url.search.slice(1, url.search.length);
+    let paramArray = url.search.slice(1, url.search.length).split("&");
+    paramArray.forEach(item => {
+      let tmp = item.split("=");
+      queryParams.set(tmp[0], tmp[1]);
+    });
+    // remove the unwanted parameters and update the query string
+    REMOVE_KEYS.forEach(key => queryParams.delete(key));
+    queryParams.forEach((val, key) => {
+      if (val.length !== 0 && key.length !== 0) {
+        safeQueryParam = safeQueryParam + key + "=" + val + "&";
+      }
+    });
+    url.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    url.hash = "";
+
+    // remove extra slashes in pathname and construct back a safe pathname
+    if (url.pathname) {
+      let pathItems = url.pathname.split("/");
+      let safePathItems = pathItems.filter(item => item.length !== 0);
+      let safePath = "/";
+      url.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url doesn't have a protocol, use https
+    if (!url.protocol) {
+      url.protocol = "https:";
+    }
+
+    return url.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createKey(url) {
```

Comment:
After discussing this with @oyiptong I agree that we should use the following layers:

1) unmodified URL (this is what is actually stored in the history and what will be loaded if the user clicks on this in their history or activity stream)

2) sanitized URL (this has some query params and basic auth scrubbed, path is normalized, includes protocol, and is what will be sent to the embedly proxy)

3) cache key (this is the sanitized URL except with the protocol dropped) this is used just for looking the URL metadata up in our local simplestorage cache

The proxy will perform no URL transformations and will query and cache URLs exactly as they are sent to it.  This way we don't wind up with a confusing situation where a URL is being transformed or miscached and we're not sure where that transformation is taking place.  If we only transform URLs in the Firefox client, then we can be sure transformation errors are handled by that code base.

The case raised by @pdehaan (the case where HTTP is disallowed, HTTPS is allowed, and we have mistakenly cached the HTTP empty dataset under the same key as the HTTPS URL) should be prevented in the client because only URLs which the browser is able to establish a successful connection to will appear in the history, so we should not be mistakenly caching data for URLs which contain no metadata.